policy-controller/k8s/index/src/outbound/index.rs

```diff
@@ -789,32 +945,64 @@ impl ServiceRoutes {
     }
 
     fn delete(&mut self, gknn: &GroupKindNamespaceName) {
-        for watch in self.watches_by_ns.values_mut() {
-            watch.routes.remove(gknn);
-            watch.send_if_modified();
-        }
+        self.watches_by_ns.values_mut().for_each(|watch| {
+            watch.remove_route(gknn);
+        });
     }
 }
 
 impl RoutesWatch {
     fn send_if_modified(&mut self) {
         self.watch.send_if_modified(|policy| {
             let mut modified = false;
-            if self.routes != policy.http_routes {
-                policy.http_routes = self.routes.clone();
+
+            if self.routes != policy.routes {
+                policy.routes = self.routes.clone();
                 modified = true;
             }
+
             if self.opaque != policy.opaque {
                 policy.opaque = self.opaque;
                 modified = true;
             }
+
             if self.accrual != policy.accrual {
                 policy.accrual = self.accrual;
                 modified = true;
             }
+
             modified
         });
     }
+
+    fn insert_route(&mut self, gknn: GroupKindNamespaceName, route: TypedOutboundRoute) {
+        match &mut self.routes {
+            Some(routes) => {
+                routes
+                    .insert(gknn, route)
+                    .map_err(log_failed_route_insertion)
+                    .transpose();
+            }
+            None => {
+                let mut routes = default_collection_for(&route);
+                routes
+                    .insert(gknn, route)
+                    .map_err(log_failed_route_insertion)
+                    .transpose();
+                self.routes = Some(routes);
+            }
+        };
+
+        self.send_if_modified();
+    }
+
+    fn remove_route(&mut self, gknn: &GroupKindNamespaceName) {
+        if let Some(routes) = &mut self.routes {
+            routes.remove(gknn);
```

Comment:
I think we need to set this back to None if it becomes empty, for the same reason as above.

Reply:
Applied